onnx/defs/math/defs.cc

```diff
@@ -1937,42 +1944,61 @@ bool BuildContextDependentFunctionBody(
       }
     }
   } else {
+    
     body.push_back(
         {{"const_ignore_index"},
          "Constant",
          {},
          {MakeAttribute(
              "value",
              ToDimensionOneTensor(ctx.getAttribute("ignore_index")->i()))}});
+
+    body.push_back({{"mask"}, "Equal", {"expanded_target", "const_ignore_index"}});
+
+    body.push_back({{"transform_targets"}, "Where", {"mask", "const_zero", "expanded_target"}});
+
+    body.push_back({{"input_gather_element"}, "GatherElements", {"input", "transform_targets"}, {MakeAttribute("axis", (int64_t)1)}});
+  
     body.push_back({{"const_zero_float"},
-                    "Constant",
-                    {},
-                    {MakeAttribute("value", ToDimensionOneFloatTensor(0.0f))}});
-    if (!ctx.hasInput(2)) {
-      body.push_back({{"const_two"},
                 "Constant",
                 {},
-                {MakeAttribute("value", ToDimensionOneTensor(2))}});
-      body.push_back({{"input_shape"}, "Shape", {"input"}});
-      body.push_back({{"input_class"},
-                      "Slice",
-                      {"input_shape", "const_one", "const_two"}});
-      body.push_back({{"const_weights_ones"},
+                {MakeAttribute("value", ToDimensionOneFloatTensor(0.0f))}});
+
+    body.push_back(
+        {{"input_gather_element_transform"}, "Where", {"mask", "const_zero_float", "input_gather_element"}});
+
+    body.push_back({{"loss_NCdd"}, "Neg", {"input_gather_element_transform"}});
+
+    body.push_back({{"loss_N1dd"},
+                    "Slice",
+                    {"loss_NCdd", "const_zero", "const_one", "const_one"}});
+
+    if (!ctx.hasInput(2)) {
+      body.push_back({{"squeeze_mask"},
+                "Squeeze",
+                {"mask"},
+                {MakeAttribute("axes", std::vector<int64_t>({1}))}});
+
+      body.push_back({{"target_shape"}, "Shape", {"target"}});
+      body.push_back({{"const_weights_ones_float"},
                       "ConstantOfShape",
-                      {"input_class"},
+                      {"target_shape"},
                       {MakeAttribute("value", ToDimensionOneFloatTensor(1))}});
-      body.push_back(
-          {{"weights_default"},
-           "ScatterElements",
-           {"const_weights_ones", "const_ignore_index", "const_zero_float"}});
-      body.push_back(
-          {{"weight_gather"}, "Gather", {"weights_default", "target"}});
+
+      body.push_back({{"weight_gather"}, "Where", {"squeeze_mask", "const_zero_float", "const_weights_ones_float"}});
```

Comment:
Minor nit: Can't we use "const_one_float" instead of "const_weights_ones_float"? Doen's "squeeze_mask" already have the desired shape? Just wondering.

Reply:
Good point, let me try.

---
In reply to: [403821108](https://github.com/onnx/onnx/pull/2703#discussion_r403821108) [](ancestors = 403821108)

Reply:
This works. Let me push the commit.

---
In reply to: [403821282](https://github.com/onnx/onnx/pull/2703#discussion_r403821282) [](ancestors = 403821282,403821108)